File: .trae/skills/umg-portfolio/evolve/scripts/inject-structure.mjs
import { readFileSync, writeFileSync, readdirSync, statSync, existsSync, mkdirSync } from 'fs';
import { join, relative } from 'path';
import { fileURLToPath } from 'url';

const __dirname = fileURLToPath(new URL('.', import.meta.url));
const ROOT = join(__dirname, '..', '..');
const HARNESSRC = join(ROOT, '.harnessrc.json');

function loadConfig() {
  if (!existsSync(HARNESSRC)) throw new Error('.harnessrc.json not found');
  return JSON.parse(readFileSync(HARNESSRC, 'utf-8'));
}

const config = loadConfig();
const PORTFOLIO = join(ROOT, config.projectRoot);
const PROJECT_DOC = join(ROOT, config.projectDocPath);

console.log('=== Project Structure Injection ===\n');

function ensureDir(dir) {
  if (!existsSync(dir)) { try { mkdirSync(dir, { recursive: true }); } catch {} }
}

function scanFiles(dir, exts, baseDir) {
  const results = [];
  try {
    const entries = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.name.startsWith('.') || entry.name === 'node_modules' || entry.name === '.next' || entry.name === 'dist') continue;
      const full = join(dir, entry.name);
      if (entry.isDirectory()) results.push(...scanFiles(full, exts, baseDir));
      else if (exts.some(e => entry.name.endsWith(e))) results.push({ path: full, rel: relative(baseDir || dir, full).replace(/\\/g, '/') });
    }
  } catch {}
  return results;
}

function autoBlock(name, content) {
  return `<!-- AUTO-START:${name} -->\n${content}\n<!-- AUTO-END:${name} -->`;
}

function replaceAutoBlock(filePath, blockName, newContent) {
  const dir = filePath.split('/').slice(0, -1).join('/');
  ensureDir(dir);
  let content = '';
  if (existsSync(filePath)) content = readFileSync(filePath, 'utf-8');
  else content = `# ${blockName}\n\n${autoBlock(blockName, '')}\n`;
  const startTag = `<!-- AUTO-START:${blockName} -->`;
  const endTag = `<!-- AUTO-END:${blockName} -->`;
  const startIdx = content.indexOf(startTag);
  const endIdx = content.indexOf(endTag);
  if (startIdx !== -1 && endIdx !== -1) {
    writeFileSync(filePath, content.slice(0, startIdx + startTag.length) + '\n' + newContent + '\n' + content.slice(endIdx));
  } else {
    writeFileSync(filePath, content.trimEnd() + '\n\n' + autoBlock(blockName, newContent) + '\n');
  }
  return true;
}

ensureDir(PROJECT_DOC);

const pkgPath = join(PORTFOLIO, 'package.json');
const pkg = existsSync(pkgPath) ? JSON.parse(readFileSync(pkgPath, 'utf-8')) : {};

const deps = Object.entries(pkg.dependencies || {}).map(([k, v]) => `${k}: ${v}`).sort();
const devDeps = Object.entries(pkg.devDependencies || {}).map(([k, v]) => `${k}: ${v}`).sort();

const nextVer = pkg.dependencies?.next || 'unknown';
const reactVer = pkg.dependencies?.react || 'unknown';
const fmVer = pkg.dependencies?.['framer-motion'] || 'unknown';
const twVer = pkg.devDependencies?.tailwindcss || 'unknown';
const tsVer = pkg.devDependencies?.typescript || 'unknown';

replaceAutoBlock(join(PROJECT_DOC, 'tech-stack.md'), 'tech-stack',
`| 项目 | 值 |
|------|-----|
| **名称** | ${pkg.name || 'unknown'} |
| **版本** | ${pkg.version || 'unknown'} |
| **框架** | Next.js ${nextVer} |
| **UI** | React ${reactVer} |
| **动画** | Framer Motion ${fmVer} |
| **样式** | Tailwind CSS v${twVer} (v4, @theme inline) |
| **语言** | TypeScript ^${tsVer} |
| **3D** | Three.js + @react-three/fiber + drei |
| **其他** | GSAP (@gsap/react), Lenis (smooth scroll), Lucide (icons), Prisma |

## Dependencies (${deps.length})

${deps.map(d => `- ${d}`).join('\n')}

## DevDependencies (${devDeps.length})

${devDeps.map(d => `- ${d}`).join('\n')}`);

const allFiles = scanFiles(PORTFOLIO, ['.tsx', '.ts', '.css', '.json'], PORTFOLIO);
let treeLines = [];
const processedDirs = new Set();

function buildTree(dir, prefix, baseDir) {
  try {
    const entries = readdirSync(dir, { withFileTypes: true }).filter(e => !e.name.startsWith('.') && e.name !== 'node_modules' && e.name !== '.next' && e.name !== 'dist').sort((a, b) => {
      if (a.isDirectory() === b.isDirectory()) return a.name.localeCompare(b.name);
      return a.isDirectory() ? -1 : 1;
    });
    for (let i = 0; i < entries.length; i++) {
      const e = entries[i];
      const isLast = i === entries.length - 1;
      const connector = isLast ? '└── ' : '├── ';
      const rel = relative(baseDir, join(dir, e.name)).replace(/\\/g, '/');
      if (e.isDirectory()) {
        treeLines.push(prefix + connector + '**' + e.name + '/**');
        buildTree(join(dir, e.name), prefix + (isLast ? '    ' : '│   '), baseDir);
      } else {
        treeLines.push(prefix + connector + rel);
      }
    }
  } catch {}
}

buildTree(PORTFOLIO, '', PORTFOLIO);

replaceAutoBlock(join(PROJECT_DOC, 'structure.md'), 'structure',
`\`\`
portfolio/
${treeLines.join('\n')}
\`\``);

const tsxFiles = scanFiles(join(PORTFOLIO, 'app'), ['.tsx'], PORTFOLIO);
const compRegex = /export\s+(default\s+)?function\s+(\w+)|export\s+(default\s+)?(?:const|var|let)\s+(\w+)\s*[\s=<]\s*(?:\(|function)/g;
const importLocalRegex = /from\s+['"]\.\.?\/([^'"]+)['"]/g;
const useClientRegex = ['use client'];

const components = [];
for (const f of tsxFiles) {
  try {
    const content = readFileSync(f.path, 'utf-8');
    const lines = content.split('\n');
    const hasUseClient = lines[0]?.trim() === "'use client'";
    const localImports = [];
    let match;
    while ((match = importLocalRegex.exec(content)) !== null) {
      localImports.push(match[1]);
    }
    const compMatches = [...content.matchAll(compRegex)];
    for (const m of compMatches) {
      const name = m[2] || m[4];
      components.push({ name, file: f.rel, lines: lines.length, client: hasUseClient, imports: localImports });
    }
  } catch {}
}

components.sort((a, b) => a.file.localeCompare(b.file));

const compTable = components.map(c =>
  `| ${c.name} | \`${c.file}\` | ${c.lines}L | ${c.client ? '✅ client' : 'server'} | ${c.imports.slice(0, 4).join(', ') || '—'} |`).join('\n');

replaceAutoBlock(join(PROJECT_DOC, 'components.md'), 'components',
`共 ${components.length} 个组件导出, 分布在 ${tsxFiles.length} 个 .tsx 文件中

| 组件名 | 文件 | 行数 | Client? | 本地依赖 |
|--------|------|------|---------|----------|
${compTable}`);

const keyPaths = [
  { label: '入口', path: 'app/page.tsx' },
  { label: '布局', path: 'app/layout.tsx' },
  { label: '内容提供者', path: 'app/ContentProvider.tsx' },
  { label: '全局样式', path: 'app/globals.css' },
  { label: '内容配置', path: 'app/config/content.ts' },
  { label: '编辑Schema', path: 'app/edit/schema.tsx' },
  { label: '编辑页面', path: 'app/edit/page.tsx' },
  { label: '保存API', path: 'app/api/save-content/route.ts' },
  { label: 'Next配置', path: 'next.config.ts' },
  { label: '包管理', path: 'package.json' },
];

const sections = [
  { name: 'Hero', path: 'app/sections/Hero.tsx' },
  { name: 'About', path: 'app/sections/About.tsx' },
  { name: 'Projects', path: 'app/sections/Projects.tsx' },
  { name: 'Skills', path: 'app/sections/Skills.tsx' },
  { name: 'Experience', path: 'app/sections/Experience.tsx' },
  { name: 'Contact', path: 'app/sections/Contact.tsx' },
  { name: 'Footer', path: 'app/sections/Footer.tsx' },
];

const sharedComponents = [
  { name: 'ParticleField', path: 'app/components/ParticleField.tsx' },
  { name: 'ProjectCard', path: 'app/components/ProjectCard.tsx' },
  { name: 'NeonCard', path: 'app/components/NeonCard.tsx' },
  { name: 'GlitchText', path: 'app/components/GlitchText.tsx' },
  { name: 'GlobalAnimatedBackground', path: 'app/components/GlobalAnimatedBackground.tsx' },
  { name: 'ParticleBackground', path: 'app/components/ParticleBackground.tsx' },
  { name: 'ScrollIndicator', path: 'app/components/ScrollIndicator.tsx' },
];

const hooks = [
  { name: 'useMousePosition', path: 'app/hooks/useMousePosition.ts' },
  { name: 'useSmoothScroll', path: 'app/hooks/useSmoothScroll.ts' },
];

const editComponents = [
  { name: 'DynamicForm', path: 'app/edit/components/DynamicForm.tsx' },
  { name: 'ArrayInput', path: 'app/edit/components/FormFields/ArrayInput.tsx' },
  { name: 'ObjectInput', path: 'app/edit/components/FormFields/ObjectInput.tsx' },
  { name: 'TextInput', path: 'app/edit/components/FormFields/TextInput.tsx' },
  { name: 'NumberInput', path: 'app/edit/components/FormFields/NumberInput.tsx' },
  { name: 'TextArea', path: 'app/edit/components/FormFields/TextArea.tsx' },
];

replaceAutoBlock(join(PROJECT_DOC, 'file-reference.md'), 'file-reference',
`### 核心
${keyPaths.map(p => `- **${p.label}**: \`portfolio/${p.path}\``).join('\n')}

### Sections (7个)
${sections.map(s => `- **${s.name}**: \`portfolio/${s.path}\``).join('\n')}

### 共享 Components (7个)
${sharedComponents.map(c => `- **${c.name}**: \`portfolio/${c.path}\``).join('\n')}

### Hooks (2个)
${hooks.map(h => `- **${h.name}**: \`portfolio/${h.path}\``).join('\n')}

### Edit 组件 (6个)
${editComponents.map(c => `- **${c.name}**: \`portfolio/${c.path}\``).join('\n')}`);

const cssPath = join(PORTFOLIO, 'app', 'globals.css');
let cssInfo = '';
if (existsSync(cssPath)) {
  const css = readFileSync(cssPath, 'utf-8');
  const varRegex = /--([\w-]+):\s*([^;]+);/g;
  const vars = [];
  let m;
  while ((m = varRegex.exec(css)) !== null) vars.push(`--${m[1]}: ${m[2]}`);
  const fontImport = css.match(/@import url\(['"]([^'"]+)['"]\)/)?.[1] || '';
  const hasTailwind = css.includes('@import "tailwindcss"');
  const hasThemeInline = css.includes('@theme inline');
  cssInfo = `- **字体**: ${fontImport || '无外部字体'}
- **Tailwind**: ${hasTailwind ? '✅ v4 (@import)' : '❌'}
- **@theme inline**: ${hasThemeInline ? '✅ (CSS变量→设计令牌)' : '❌'}
- **CSS 变量** (${vars.length} 个):

\`\`\`css
${vars.join('\n')}
\`\`\``;
} else {
  cssInfo = '*globals.css 未找到*';
}

replaceAutoBlock(join(PROJECT_DOC, 'css-classes.md'), 'css-classes', cssInfo);

const providerPath = join(PORTFOLIO, 'app', 'ContentProvider.tsx');
let dataFlow = '';
if (existsSync(providerPath)) {
  const provider = readFileSync(providerPath, 'utf-8');
  const contentKeySet = new Set();
  [...provider.matchAll(/\b(hero|about|projects|skills|contact|footer)Content\b/g)].forEach(m => contentKeySet.add(m[1]));
  const contentKeys = [...contentKeySet];

  const consumers = [];
  for (const f of tsxFiles) {
    try {
      const c = readFileSync(f.path, 'utf-8');
      if (c.includes('useContent()')) {
        const compName = f.rel.split('/').pop()?.replace('.tsx', '') || f.rel;
        consumers.push(compName);
      }
    } catch {}
  }

  dataFlow = `\`\`\`
content.json (静态默认值)
    ↓ fetch() + localStorage merge
ContentProvider.tsx (React Context)
    ↓ useState(defaultContent) → useEffect(远程覆盖+本地存储合并)
    ↓
useContent() hook → 返回 EditedContent 对象
    ↓
Consumers (${consumers.length}):
  ${consumers.map(c => `  └─ ${c}()`).join('\n')}
\`\`

**数据结构**:
${contentKeys.map(k => `- **${k}** — 定义于 config/content.ts, 类型见 edit/schema.tsx`).join('\n')}

**写入路径**:
- API: POST \`api/save-content/route.ts\` → 写入 public/content.json
- 客户端: ContentProvider fetch(\`/content.json\`) + localStorage 读写
- 编辑器: edit/page.tsx → DynamicForm → schema.tsx 验证 → save-content API`;
} else {
  dataFlow = '*ContentProvider 未找到*';
}

replaceAutoBlock(join(PROJECT_DOC, 'data-flow.md'), 'data-flow', dataFlow);

const nextConfigPath = join(PORTFOLIO, 'next.config.ts');
let buildInfo = '';
if (existsSync(nextConfigPath)) {
  const nc = readFileSync(nextConfigPath, 'utf-8');
  const outputMatch = nc.match(/output:\s*['"]([^'"]*)['"]/);
  const distMatch = nc.match(/distDir:\s*['"]([^'"]*)['"]/);
  const basePathMatch = nc.match(/basePath:\s*[^,]+?['"]\/([^'"]*)['"]/);
  const scriptLines = Object.entries(pkg.scripts || {}).map(([k, v]) => `- \`${k}\`: ${v}`).join('\n');
buildInfo = `| 配置项 | 值 | 说明 |
|--------|-----|------|
| **output** | \`${outputMatch?.[1] || '?'}\` | ${outputMatch?.[1] === 'export' ? '静态导出模式 ✅' : 'SSR 模式'} |
| **distDir** | \`${distMatch?.[1] || '?'}\` | 构建输出目录 |
| **basePath** | \`${basePathMatch?.[1] ? '/' + basePathMatch?.[1] : '(dev: 空)'}\` | 生产环境前缀 |
| **images.unoptimized** | \`${nc.includes('unoptimized: true') ? 'true ✅' : 'false'}\` | 禁用 Image 优化 (静态导出需要) |

**Scripts**:
${scriptLines}`;
} else {
  buildInfo = '*next.config.ts 未找到*';
}

replaceAutoBlock(join(PROJECT_DOC, 'build-config.md'), 'build-config', buildInfo);

console.log(`Project Doc generated at: ${PROJECT_DOC}`);
console.log(`Files: tech-stack / structure / components / file-reference / css-classes / data-flow / build-config`);
console.log(`Components found: ${components.length}`);
console.log(`Total source files scanned: ${allFiles.length}`);

File: .trae/skills/umg-portfolio/scripts/inject_structure.py
#!/usr/bin/env python3
"""
Inject Structure - Generate ProjectDoc files from codebase analysis.
Creates structure.md, components.md, file-reference.md, tech-stack.md, etc.
"""
import os
import re
import json
from pathlib import Path
from datetime import datetime
from collections import defaultdict

SCRIPT_DIR = Path(__file__).parent
SKILL_DIR = SCRIPT_DIR.parent
PROJECT_ROOT = SKILL_DIR.parent.parent.parent
PORTFOLIO_DIR = PROJECT_ROOT / "portfolio"
PROJECT_DOC_DIR = PROJECT_ROOT / "ProjectDoc"

EXTENSIONS = [".tsx", ".ts", ".css"]
EXCLUDE_DIRS = {"node_modules", ".next", "dist", ".git", "out"}


def scan_files(directory: Path, exts: list[str]) -> list[Path]:
    """Recursively scan directory for source files."""
    results = []
    try:
        for entry in directory.iterdir():
            if entry.name.startswith(".") or entry.name in EXCLUDE_DIRS:
                continue
            if entry.is_dir():
                results.extend(scan_files(entry, exts))
            elif entry.suffix in exts:
                results.append(entry)
    except PermissionError:
        pass
    return results


def count_lines(file_path: Path) -> int:
    """Count lines in a file."""
    try:
        return len(file_path.read_text(encoding="utf-8").split("\n"))
    except Exception:
        return 0


def has_use_client(file_path: Path) -> bool:
    """Check if file has 'use client' directive."""
    try:
        content = file_path.read_text(encoding="utf-8")
        return content.strip().startswith("'use client'") or content.strip().startswith('"use client"')
    except Exception:
        return False


def extract_imports(content: str) -> tuple[list[str], list[str]]:
    """Extract local and external imports."""
    local = []
    external = []
    import_pattern = re.compile(r"import\s+.*?from\s+['\"]([^'\"]+)['\"]")
    for match in import_pattern.finditer(content):
        module = match.group(1)
        if module.startswith(".") or module.startswith("@/"):
            local.append(module)
        else:
            external.append(module)
    return local, external


def extract_component_name(file_path: Path) -> str:
    """Extract component name from file."""
    return file_path.stem


def analyze_components(files: list[Path]) -> list[dict]:
    """Analyze component files."""
    components = []
    for f in files:
        try:
            content = f.read_text(encoding="utf-8")
            local_deps, ext_deps = extract_imports(content)
            components.append({
                "name": extract_component_name(f),
                "path": str(f.relative_to(PORTFOLIO_DIR)),
                "lines": count_lines(f),
                "client": has_use_client(f),
                "local_deps": local_deps,
                "ext_deps": ext_deps
            })
        except Exception:
            pass
    return components


def generate_structure_md() -> str:
    """Generate structure.md content."""
    lines = ["# Project Structure\n"]
    lines.append(f"Generated: {datetime.now().isoformat()[:19]}\n")
    lines.append("```\n")
    
    def walk_dir(directory: Path, prefix: str = ""):
        try:
            entries = sorted(directory.iterdir(), key=lambda x: (not x.is_dir(), x.name))
            for i, entry in enumerate(entries):
                if entry.name.startswith(".") or entry.name in EXCLUDE_DIRS:
                    continue
                is_last = i == len(entries) - 1
                connector = "└── " if is_last else "├── "
                lines.append(f"{prefix}{connector}{entry.name}")
                if entry.is_dir():
                    new_prefix = prefix + ("    " if is_last else "│   ")
                    walk_dir(entry, new_prefix)
        except PermissionError:
            pass
    
    walk_dir(PORTFOLIO_DIR)
    lines.append("```\n")
    return "\n".join(lines)


def generate_components_md(components: list[dict]) -> str:
    """Generate components.md content."""
    lines = ["# Components\n"]
    lines.append(f"Total: {len(components)} components\n")
    lines.append("| Name | Path | Lines | Client | Local Deps |")
    lines.append("|------|------|-------|--------|------------|")
    for c in components:
        client = "✓" if c["client"] else ""
        local = ", ".join(c["local_deps"][:3]) + ("..." if len(c["local_deps"]) > 3 else "")
        lines.append(f"| {c['name']} | {c['path']} | {c['lines']} | {client} | {local} |")
    return "\n".join(lines)


def generate_tech_stack_md() -> str:
    """Generate tech-stack.md content."""
    package_json = PORTFOLIO_DIR / "package.json"
    if not package_json.exists():
        return "# Tech Stack\n\npackage.json not found"
    
    try:
        data = json.loads(package_json.read_text(encoding="utf-8"))
    except Exception:
        return "# Tech Stack\n\nFailed to parse package.json"
    
    lines = ["# Tech Stack\n"]
    lines.append(f"Generated: {datetime.now().isoformat()[:19]}\n")
    
    deps = data.get("dependencies", {})
    dev_deps = data.get("devDependencies", {})
    
    lines.append("## Dependencies\n")
    for name, version in sorted(deps.items()):
        lines.append(f"- **{name}**: {version}")
    
    lines.append("\n## Dev Dependencies\n")
    for name, version in sorted(dev_deps.items()):
        lines.append(f"- **{name}**: {version}")
    
    return "\n".join(lines)


def generate_file_reference_md(components: list[dict]) -> str:
    """Generate file-reference.md content."""
    lines = ["# File Reference\n"]
    lines.append(f"Generated: {datetime.now().isoformat()[:19]}\n")
    
    sections = [c for c in components if "sections" in c["path"]]
    comps = [c for c in components if "components" in c["path"] and "sections" not in c["path"]]
    
    lines.append("## Sections\n")
    for c in sections:
        lines.append(f"- [{c['name']}](../portfolio/{c['path']}) ({c['lines']} lines)")
    
    lines.append("\n## Components\n")
    for c in comps:
        lines.append(f"- [{c['name']}](../portfolio/{c['path']}) ({c['lines']} lines)")
    
    return "\n".join(lines)


def main():
    if not PORTFOLIO_DIR.exists():
        print(f"Portfolio directory not found: {PORTFOLIO_DIR}")
        return
    
    PROJECT_DOC_DIR.mkdir(parents=True, exist_ok=True)
    
    tsx_files = scan_files(PORTFOLIO_DIR / "app", [".tsx"])
    components = analyze_components(tsx_files)
    
    (PROJECT_DOC_DIR / "structure.md").write_text(generate_structure_md(), encoding="utf-8")
    print(f"Generated: structure.md")
    
    (PROJECT_DOC_DIR / "components.md").write_text(generate_components_md(components), encoding="utf-8")
    print(f"Generated: components.md ({len(components)} components)")
    
    (PROJECT_DOC_DIR / "tech-stack.md").write_text(generate_tech_stack_md(), encoding="utf-8")
    print(f"Generated: tech-stack.md")
    
    (PROJECT_DOC_DIR / "file-reference.md").write_text(generate_file_reference_md(components), encoding="utf-8")
    print(f"Generated: file-reference.md")
    
    print(f"\n=== Structure Injection Complete ===")
    print(f"ProjectDoc: {PROJECT_DOC_DIR}")


if __name__ == "__main__":
    main()
